### PR TITLE
[Editorial] Meet In The Middle On Trees

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [ ] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: solutions/gold/kattis-playlist.mdx [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		}
+	}
+
+	// Compute the 4-length chains starting from node i
+	for (int i = 0; i < n; i++) {
+		dfs({}, i, 0, 0);
+		dfs({}, i, 0, 1);
+	}
+
+	// Optimization trick
```
Comment (envyaims) [inline solutions/gold/kattis-playlist.mdx:103]: whats the trick

Comment (brebenelmihnea) [reply solutions/gold/kattis-playlist.mdx:103]: It gets TLE without

Comment (brebenelmihnea) [reply solutions/gold/kattis-playlist.mdx:103]: Nevermind, it gets AC anyway

Comment (brebenelmihnea) [reply solutions/gold/kattis-playlist.mdx]: Removed

File: solutions/gold/kattis-playlist.mdx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+						s[k] = 0;
+						odd ^= 1;
+					}
+				}
+				sort(s.begin(), s.end());
+				// Inclusion-Exclusion on sets
+				cnt += (1 - 2 * odd) * m[getid(s)];
+			}
+			if (cnt) {
+				for (State t : states[0][i]) {
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-playlist.mdx:136]: when you're looping thru the states, do they have to be copies? or can they be const refs?

Comment (brebenelmihnea) [reply solutions/gold/kattis-playlist.mdx:136]: No they don't have.

Comment (brebenelmihnea) [reply solutions/gold/kattis-playlist.mdx]: Done, I made all const refs

File: solutions/gold/kattis-playlist.mdx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+				auto s = state;
+				for (int k = 0; k < 4; k++) {
+					if (j & (1 << k)) { s[k] = 0; }
+				}
+				sort(s.begin(), s.end());
+				m[getid(s)]++;
+			}
+		}
+		for (const State &state : states[1][i]) {
+			int cnt = (int)states[0][i].size();
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-playlist.mdx:121]: better name

Comment (brebenelmihnea) [reply solutions/gold/kattis-playlist.mdx]: Done

File: solutions/gold/kattis-playlist.mdx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+				auto s = state;
+				for (int k = 0; k < 4; k++) {
+					if (j & (1 << k)) { s[k] = 0; }
+				}
+				sort(s.begin(), s.end());
+				m[getid(s)]++;
+			}
+		}
+		for (const State &state_path : states[1][i]) {
+			int cnt_paths = (int)states[0][i].size();
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-playlist.mdx:121]: rename to `path_count`

File: solutions/gold/kattis-playlist.mdx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	for (int i = 0; i < n; i++) {
+		dfs({}, i, 0, 0);
+		dfs({}, i, 0, 1);
+	}
+
+	for (int i = 0; i < n; i++) {
+		unordered_map<long long, int> m;
+		for (State state : states[0][i]) {
+			for (int j = 0; j < 16; j++) {
+				auto s = state;
```
Comment (SansPapyrus683) [inline solutions/gold/kattis-playlist.mdx:112]:
```suggestion
				State s = state;
```

File: solutions/gold/kattis-playlist.mdx (hunk 4 above)
Comment (brebenelmihnea) [reply solutions/gold/kattis-playlist.mdx:121]: Done

File: solutions/gold/kattis-playlist.mdx (hunk 5 above)
Comment (brebenelmihnea) [reply solutions/gold/kattis-playlist.mdx]: Done

**Review by SansPapyrus683 — APPROVED** [review]
ss of ac?

**Comment by brebenelmihnea** [comment]
AC proof

https://open.kattis.com/submissions/13817355

image

**Comment by SansPapyrus683** [comment]
get envy to approave and ill merge

**Review by envyaims — APPROVED** [review]

**Pull request merged**